### PR TITLE
CP-23536: Add event loop listening to network changes and updating the firewall for the NBD server.

### PR DESCRIPTION
**Pull request by gaborigloi**

### CODE REVIEW

**Comment by thomassa** [comment]
The new files need copyright headers.

**Comment by gaborigloi** [comment]
I have tested this with and without a bond, and it worked fine. I used the new `test-nbd-network-config` test in https://github.com/gaborigloi/cbt-tests

**Comment by gaborigloi** [comment]
There is an edge case not covered by this event loop: when there are NBD-enabled networks, and then all the networks in the database get destroyed, and xapi stops before the event loop could process the event. When it restarts, the event loop will not ban NBD on all interfaces at startup, because there are not network objects at all, so in this case the first `Event.from` invocation will not return immediately with the snapshots of all the networks. So this means that the previous interfaces that were allowed before the network objects got destroyed will still remain allowed, until the next event on the network objects. Not sure how important this edge case is though.
Actually the version of the untit tests before my previous commit caught this edge case.
One way to fix this would be to add an unconditional firewall update to the startup of the event loop.

File: ocaml/xapi/network_event_loop.ml [hunk 1]
```diff
@@ -0,0 +1,92 @@
+module D=Debug.Make(struct let name="network_event_loop" end)
+
+module StringMap = Map.Make(struct type t = string let compare = compare end)
+
+type 'a update = Deleted | Snapshot of 'a
```
Comment (jonludlam) [inline ocaml/xapi/network_event_loop.ml]: I don't think you're using this any more

File: ocaml/xapi/network_event_loop.ml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  let wait_for_network_change ~token =
+    let from =
+      Helpers.call_api_functions ~__context
+        (fun rpc session_id ->
+           Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout |> Event_types.event_from_of_rpc)
+        ~test_fn:(fun () -> Xapi_event.from ~__context ~classes ~token ~timeout |> Event_types.parse_event_from)
+    in
+    let events = from.Event_types.events in
+    List.iter (fun event -> D.debug "got event %s" (Event_types.string_of_event event)) events;
```
Comment (jonludlam) [inline ocaml/xapi/network_event_loop.ml]: Probably don't want to log every single event here

File: ocaml/xapi/network_event_loop.ml (hunk 1 above)
Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: Oops thanks, I forgot to delete that.

File: ocaml/xapi/network_event_loop.ml (hunk 2 above)
Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: Hmm ok, I'll remove it. My thinking was that this should not be triggered oftern, because we're just using events here as a notification mechanism, and query the DB. So if this is triggered too often, that will lead to many database queries and inefficiency.

**Comment by gaborigloi** [comment]
I have verified that this also works when I connect to a slave, using the `--hostname` option of [cbt-tests](https://github.com/gaborigloi/cbt-tests), which connects to a VDI that is on a local SR only connected to the given host.

File: ocaml/xapi/network_event_loop.ml (hunk 1 above)
Comment (thomassa) [reply ocaml/xapi/network_event_loop.ml]: Using it on line 41 of the first commit, but perhaps it goes away later...

File: ocaml/xapi/network_event_loop.ml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+     interfaces. *)
+  let allowed_interfaces = [] in
+
+  let api_timeout = 60. in
+  let timeout = 30. +. api_timeout +. !Db_globs.master_connection_reset_timeout in
+
+  let rec update_networks ~token ~networks =
+
+    let most_recent_state events =
+      let events_from_newest_to_oldest =
```
Comment (thomassa) [inline ocaml/xapi/network_event_loop.ml]: Are you sure we can't rely on them being sorted already?

File: ocaml/xapi/network_event_loop.ml (hunk 2 above)
Comment (thomassa) [reply ocaml/xapi/network_event_loop.ml]: Right, if events on the network class are common, we probably have bigger problems.
I imagine that these log-lines are not that likely to be useful for diagnosing problems in real deployments though.

File: ocaml/xapi/network_event_loop.ml (hunk 3 above)
Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: This intermediate commit has been changed by a subsequent one - I though I'd just keep it in case we decide to use caching - sorry for the confusion.

Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: However, to answer your question, it's not mentioned in the XenAPI reference, and I chose not to rely on assumptions that are not part of our official interface/contract/documentation.

File: ocaml/xapi/network_event_loop.ml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  (* We keep track of the network objects in the database using this event loop. *)
+  let classes = ["network"] in
+  (* We keep track of the interfaces that we last passed to the firewall script
+     to allow NBD traffic on them. At startup, we don't know on which
+     interfaces NBD is allowed, and we always update the firewall. *)
+  let allowed_interfaces = None in
+
+  let api_timeout = 60. in
+  let timeout = 30. +. api_timeout +. !Db_globs.master_connection_reset_timeout in
```
Comment (gaborigloi) [inline ocaml/xapi/network_event_loop.ml]: I literally copied this from `Xapi_xenops.events_from_xapi`.

File: ocaml/xapi/network_event_loop.ml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  D.debug "Listening to events on network objects";
+  loop ~token:"" ~allowed_interfaces
+
+let update_firewall interfaces_allowed_for_nbd =
+  let args = "set" :: interfaces_allowed_for_nbd in
+  Forkhelpers.execute_command_get_output !Xapi_globs.nbd_firewall_config_script args |> ignore
+
+let watch_networks () =
+  Server_helpers.exec_with_new_task "watching changes in network objects" (_watch_networks ~update_firewall ~wait_after_failure_seconds:5.0)
```
Comment (gaborigloi) [inline ocaml/xapi/network_event_loop.ml]: I'm not sure what the best delay is here for `wait_after_failure_seconds`, `Xapi_xenopsd.events_from_xapi` uses 15 seconds for all failure (including `events_lost` - I don't wait in case of `events_lost` here, because that's due to slow processing of events).

File: ocaml/xapi/network_event_loop.ml (hunk 1 above)
Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: Yep. It's better to just check the whole diff.

**Comment by gaborigloi** [comment]
**Note for reviewers:** it's the easiest to check the whole diff, because I've kept commits that are changed significantly by subsequent commits, because the code is useful, in case we want to use caching in the future. But I'm happy to squash this together if that's not the case.

File: ocaml/xapi/network_event_loop.ml (hunk 3 above)
Comment (jonludlam) [reply ocaml/xapi/network_event_loop.ml]: There's no guarantee of ordering of events, but you'll get each object only once, containing the final state of the object.

**Comment by gaborigloi** [comment]
The Travis build fails because we've just done a rebase. 
Also, it seems that for some reason when using the `Debug` module from `xcp-idl` in a new thread, the logs go to both stdout and syslog :thinking: , this happens with the new unit test too.

File: ocaml/xapi/network_event_loop.ml [hunk 6]
```diff
@@ -0,0 +1,88 @@
+module D=Debug.Make(struct let name="network_event_loop" end)
+
+module StringMap = Map.Make(struct type t = string let compare = compare end)
```
Comment (robhoes) [inline ocaml/xapi/network_event_loop.ml]: I don't see this used anywhere.

**Comment by robhoes** [comment]
Does the firewall need to be updated when the IP address of an NBD interface changes? If so, we'd need to watch the PIFs as well.

**Comment by robhoes** [comment]
It may be worth rate-limiting this by adding a delay at the end of the loop. E.g. to not run through the loop more often than once every 10s. Just in case some network fields are updated a lot.

**Comment by gaborigloi** [comment]
@robhoes No, I don't think so, because we enable NBD on the `bridge` field of the `network` objects.
This event loop is "idempotent" in that it does not call the firewall updating script if the set of interfaces for which NBD should be allowed did not change. So this means that if there are irrelevant `network` changes, we will just loop on the events, but not do anything else, which should be harmless?

File: ocaml/xapi/network_event_loop.ml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        let interface_list = String.concat ", " interfaces in
+        if needs_firewall_update then begin
+          D.debug "Updating the firewall to use the following interfaces for NBD: [%s]" interface_list;
+          update_firewall interfaces
+        end else begin
+          D.debug "Not updating the firewall, because the set of interfaces to use for NBD did not change: [%s]" interface_list
+        end;
+        (token, Some interfaces)
+      with
+      | Api_errors.Server_error (code, _) as e when code = Api_errors.events_lost ->
```
Comment (robhoes) [inline ocaml/xapi/network_event_loop.ml]: @jonludlam Can `event.from` raise this error?

File: ocaml/xapi/xapi.ml [hunk 8]
```diff
@@ -913,6 +913,7 @@ let server_init() =
           (* CA-13878: make sure PBD plugging has happened before attempting to reboot any VMs *)
           "resynchronising VM state", [], (fun () -> Xapi_xenops.on_xapi_restart ~__context);
           "listening to events from xapi", [], (fun () -> if not (!noevents) then ignore (Thread.create Xapi_xenops.events_from_xapi ()));
+          "listening to events from xapi", [], (fun () -> ignore (Thread.create Network_event_loop.watch_networks ()));
```
[the remark this answers is not in the document]
Comment (gaborigloi) [reply ocaml/xapi/xapi.ml]: Yep, I'll do that, thanks. I'll still need `Server_helpers.exec_with_new_task`, because this has to have type `unit -> unit`, but I still need a `Context`.

File: ocaml/xapi/test_network_event_loop.ml [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+           ~wait_after_failure_seconds:network_event_loop_wait_after_failure_seconds
+      )
+  in
+
+  let assert_received_params msg expected =
+    Thread.delay delay;
+    match !received_params with
+    | None -> OUnit.assert_failure ("The update_firewall function was not called: " ^ msg)
+    | Some p ->
+      Ounit_comparators.StringSet.(assert_equal (of_list expected) (of_list p))
```
Comment (thomassa) [inline ocaml/xapi/test_network_event_loop.ml]: The first item has to be the same string ("set") in each list: just checking the lists contain the same values is not enough.
So maybe add a line or two to check that if the expected list has at least one value then the first values match.

File: ocaml/xapi/network_event_loop.ml (hunk 7 above)
Comment (jonludlam) [reply ocaml/xapi/network_event_loop.ml]: yes, but it's very rare - when you've missed more than 500 object deletions. It's way less common than for event.next.

File: ocaml/xapi/test_network_event_loop.ml [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+  let network1 = Test_common.make_network ~__context ~bridge:"bridge1" () in
+
+  if no_nbd_networks_at_start then begin
+    received_params := None;
+    let _ : Thread.t = start_event_loop () in
+    assert_received_params "The event loop should always update the firewall state at startup, even if no networks are allowed for NBD" []
+  end;
+
+  (* Add a new network "network1" with a PIF connected to this host *)
```
Comment (thomassa) [inline ocaml/xapi/test_network_event_loop.ml]: The comment here is wrong: the network exists already but doesn't have a PIF connected to this host.

File: ocaml/xapi/test_network_event_loop.ml [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+  (* Now wait for the event loop to reregister and continue *)
+  Thread.delay network_event_loop_wait_after_failure_seconds;
+  (* Test that transient failures in the script won't stop the event loop, and
+     that we will eventually process the event we missed *)
+  assert_received_params "we should have noticed the addition of network2 and its PIF after the transient failure" ["bridge1"; "bridge2"];
+
+  (* Add a new network "network3" with a PIF connected to another host *)
+  received_params := None;
+  let network3 = Test_common.make_network ~__context ~purpose:[`nbd] ~bridge:"bridge3" () in
+  assert_not_called "network3 has no PIF connected to this host" ();
```
Comment (thomassa) [inline ocaml/xapi/test_network_event_loop.ml]: Tiny improvement: this would be clearer as "network3 has no PIFs at all."

File: ocaml/xapi/test_network_event_loop.ml [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+  (* Remove pif3 of network3 *)
+  received_params := None;
+  Db.PIF.destroy ~__context ~self:pif3;
+  (* This was network3's only PIF connected to this host, so we should have disabled NBD on network3's bridge *)
+  assert_received_params "network3's only PIF connected to this host has been removed" ["bridge2"];
+
+  (* Remove network2 *)
+  received_params := None;
+  fail_firewall_update := true;
```
Comment (thomassa) [inline ocaml/xapi/test_network_event_loop.ml]: There's code earlier that tests the update-failure handling. Do we gain anything by doing it again here?

Comment (gaborigloi) [reply ocaml/xapi/test_network_event_loop.ml]: Yes, the comment also says what's different here: we have an empty list of interfaces, unlike in the previous case.

File: ocaml/xapi/xapi_globs.ml [hunk 13]
```diff
@@ -1075,6 +1077,7 @@ module Resources = struct
     "non-managed-pifs", non_managed_pifs, "Executed during PIF.scan to find out which NICs should not be managed by xapi";
     "update-issue", update_issue_script, "Running update-service when configuring the management interface";
     "killall", kill_process_script, "Executed to kill process";
+    "nbd-firewall-config", nbd_firewall_config_script, "Executed after NBD-related networking changes to configure the firewall for NBD";
```
Comment (thomassa) [inline ocaml/xapi/xapi_globs.ml]: This name makes it sound like a file of configuration data, but it's good to have it match the script name.
I (we) should have thought of this when I named the script in the first place...

Comment (gaborigloi) [reply ocaml/xapi/xapi_globs.ml]: So should I leave it as it is for consistency, because this is the name of the script?

File: ocaml/xapi/network_event_loop.ml [hunk 14]
```diff
@@ -82,5 +82,5 @@ let update_firewall interfaces_allowed_for_nbd =
   let args = "set" :: interfaces_allowed_for_nbd in
   Forkhelpers.execute_command_get_output !Xapi_globs.nbd_firewall_config_script args |> ignore
 
-let watch_networks () =
-  Server_helpers.exec_with_new_task "watching changes in network objects" (_watch_networks ~update_firewall ~wait_after_failure_seconds:5.0)
+let watch_networks_for_nbd_changes () =
+  Server_helpers.exec_with_new_task "watching networks for NBD-related changes" (_watch_networks_for_nbd_changes ~update_firewall ~wait_after_failure_seconds:5.0)
```
Comment (robhoes) [inline ocaml/xapi/network_event_loop.ml]: You don't need `Server_helpers.exec_with_new_task` here anymore, because the `Startup` module does that.

File: ocaml/xapi/xapi_globs.ml (hunk 13 above)
Comment (robhoes) [reply ocaml/xapi/xapi_globs.ml]: I was thinking the same... But if the script is called this way, then fine.

File: ocaml/xapi/test_network_event_loop.ml [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+         Network_event_loop._watch_networks
+           __context
+           ~update_firewall:(fun pifs ->
+               if !fail_firewall_update then failwith "Failed to update firewall";
+               received_params := Some pifs
+             )
+           ~wait_after_failure_seconds:network_event_loop_wait_after_failure_seconds
+      )
+  in
+
```
Comment (thomassa) [inline ocaml/xapi/test_network_event_loop.ml]: There's a common pattern of
```
fail_firewall_update := None;
...do setup stuff...
assert_received_params msg expected;
```
(or `assert_not_called` at the end) so I wonder about changing those two functions to be like this:
```
let assert_received_params setup msg expected =
  received_params := None;
  setup ();
  Thread.delay delay;
  ...examine received_params with OUnit.assert_something...
in
```

File: ocaml/xapi/network_event_loop.ml (hunk 14 above)
Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: I still need a `context`, and the startup module does not give me a `context` :disappointed:

File: ocaml/xapi/test_network_event_loop.ml (hunk 12 above)
Comment (thomassa) [reply ocaml/xapi/test_network_event_loop.ml]: Ah, yes, the "even if it the empty list" part of the comment. Right then.

File: ocaml/xapi/test_network_event_loop.ml (hunk 15 above)
Comment (gaborigloi) [reply ocaml/xapi/test_network_event_loop.ml]: I agree with you that the test is not ideal right now, and I'll improve it once I can think of a better way :disappointed: I wanted to test the long-running behaviour of this event loop, and also how it behaves in case of failures.

File: ocaml/xapi/valid_ref_list.ml [hunk 16]
```diff
@@ -16,3 +16,5 @@ let map f = Stdext.Listext.List.filter_map (default_on_missing_ref (fun x -> Som
 
 let flat_map f l = List.map (default_on_missing_ref f []) l |> List.flatten
 
+let filter_map f l = map f l |> filter ((<>) None) |> map Xapi_stdext_monadic.Opt.unbox
```
Comment (thomassa) [inline ocaml/xapi/valid_ref_list.ml]: For the `filter` and the second `map` we can use the versions from `List`.

Comment (gaborigloi) [reply ocaml/xapi/valid_ref_list.ml]: Good idea.

File: ocaml/xapi/test_valid_ref_list.ml [hunk 17]
```diff
@@ -69,6 +69,29 @@ let test_flat_map =
       assert_equal ["a"; "d_a"; "d"; "d_d"] (Valid_ref_list.flat_map f l)
     )
 
+let test_filter_map =
+  with_vm_list (fun __context ->
+      function
+      | [vm1; vm2; vm3; vm4] as l ->
+        let f vm =
+          let l = Db.VM.get_name_label ~__context ~self:vm in
```
Comment (thomassa) [inline ocaml/xapi/test_valid_ref_list.ml]: Looks correct, but would be more readable with `n` instead of `l` for the name-label.

File: ocaml/xapi/xapi_globs.ml (hunk 13 above)
Comment (thomassa) [reply ocaml/xapi/xapi_globs.ml]: Maybe worth changing afterwards, in a coordinated pair of pull-requests (here and in xenserver-specs).
New name might be "nbd-firewall-setter" or "nbd-firewall-tool" or "nbd-firewall-interfaces"?

File: ocaml/xapi/test_valid_ref_list.ml (hunk 17 above)
Comment (gaborigloi) [reply ocaml/xapi/test_valid_ref_list.ml]: Good idea, done.

**Comment by thomassa** [comment]
Why are the Travis builds failing?

**Comment by gaborigloi** [comment]
We've recently rebased and the new build hasn't finished yet.

**Comment by thomassa** [comment]
Looks fine now (but will need squashing).

File: ocaml/xapi/xapi_globs.ml (hunk 13 above)
Comment (gaborigloi) [reply ocaml/xapi/xapi_globs.ml]: Sounds good. We were discussing with Rob that in the future we may want to have one, unified way of changing the firewall config.

File: ocaml/xapi/network_event_loop.ml [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+                 Some network
+               else
+                 None
+            )
+            pifs
+        in
+        let interfaces = List.map (fun network -> Db.Network.get_bridge ~__context ~self:network) allowed_connected_networks in
+        let needs_firewall_update = match allowed_interfaces with
+          | Some allowed_interfaces ->
+            not (Xapi_stdext_std.Listext.List.set_equiv interfaces allowed_interfaces)
```
Comment (gaborigloi) [inline ocaml/xapi/network_event_loop.ml]: The problem with idempotency is that if people mess with the firewall config themselves, e.g. using the NBD firewall config script, than things will stop working. But I guess we need this now, because it's best not to update the firewall after every irrelevant network object change.

Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: But actually users should not change dom0's firewall. 
If xapi getting out of sync turns out to be an issue in the future, we can use the older, caching code, and we can compare the new and old snapshots to see if any relevant fields change. Then we could call the firewall unconditionally, because we'd know when we have to update it. (the code for the old, caching version of this event loop is here: https://github.com/gaborigloi/xen-api/blob/feature/CBT_network_event_caching/ocaml/xapi/network_event_loop.ml )

Comment (gaborigloi) [reply ocaml/xapi/network_event_loop.ml]: There is an older version of this code, which does caching, and we can compare the new and old snapshots to see if any relevant fields change. Then we'd know when we have to update update the firewall, and would not have to make DB queries. We can revert to this version if the current one turns out to be slow due to the DB queries. (the code is here: https://github.com/gaborigloi/xen-api/blob/feature/CBT_network_event_caching/ocaml/xapi/network_event_loop.ml )
